Documentation/proposals/20250122-blackbox-crd.md

```diff
@@ -0,0 +1,148 @@
+
+* **Other docs:**
+  * n/a
+
+This document proposes the creation of a Blackbox Operator and associated CRDs, similar to Prometheus-Operator, to manage blackbox-exporter instances and introduce a Module CRD for handling dynamic configuration of blackbox modules in a Kubernetes-native way.
+
+
+
+## Why
+In our daily use, we have discovered that the Probe CRD is extremely convenient for ICMP prober. The Probe CRD can define anything ICMP needed, but when we want to use HTTP Probe, we must define header/body and plenty of other parameters in blackbox config first, because HTTP is much more complex than ICMP.
+
+The situation is very similar to Alertmanager/AlertmanagerConfig that prometheus-operator does. We need a Kubernetes-native way to generate blackbox config dynamically by reading the Probe Module Config.
+
+
+## Pitfalls of the current solution
+
+The Probe CRD provides a partial solution for handling complex modules like HTTP and gRPC. However, it introduces a separate lifecycle for Blackbox management. Users must pre-define modules containing specific configurations (such as headers and request bodies) by manually updating and reloading the Blackbox configuration, before referencing these modules in their probes.
+
+## Goals
+
+- Provide a way for users to self-service adding module config for their Probe CRD
+- Consolidate the module configuration generation logic in a central point for other resources to use
+
+## Audience
+
+* Administrators providing Prometheus monitoring who want to enable self-service probe configuration for their users
+* Teams wanting to manage blackbox module configurations declaratively through Kubernetes CRDs
+* Users requiring a standardized, Kubernetes-native approach to monitor external targets
+
+## Non-Goals
+
+- This propose don't aim to replace the probe crd
+
+## How
+
+We propose creating two new Custom Resource Definitions (CRDs):
+- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances
+- The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator
+
+## Alternatives
+
+- Update blackbox configuration whenever a probe module is modified or added
+
+## Action Plan
+
+1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
+2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates
+
+## CRDs Drafg
+
+```yaml
+apiVersion: monitoring.zhipu.ai/v1alpha1
+kind: BlackboxExporter
+metadata:
+  name: blackbox-exporter-draft
+spec:
+  # Deployment mode - how the BlackboxExporter should be deployed 
+    # Or Daemonset mode
+  deployMode: Deployment
```

Comment:
is there a real need for a daemonset  mode? what would be the use case?

Reply:
+1